contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaSource.java

```diff
@@ -0,0 +1,806 @@
+    }
+
+    @Override
+    public void close() throws IOException {
+      Closeables.closeQuietly(consumer);
+    }
+  }
+
+  // Builder, Source, and Reader wrappers. Often user is only interested in Value in KafkaRecord :
+
+  /**
+   * Builder for Kafka Source where user is not interested in Kafka metadata and key for a record,
+   * but just the value.
+   */
+  public static class ValueSourceBuilder<V> {
+
+    private Builder<byte[], V> underlying;
+
+    private ValueSourceBuilder(Builder<byte[], V> underlying) {
+      this.underlying = underlying;
+    }
+
+    public ValueSourceBuilder<V> withBootstrapServers(String bootstrapServers) {
+      return new ValueSourceBuilder<V>(underlying.withBootstrapServers(bootstrapServers));
+    }
+
+    public ValueSourceBuilder<V> withTopics(Collection<String> topics) {
+      return new ValueSourceBuilder<V>(underlying.withTopics(topics));
+    }
+
+    public ValueSourceBuilder<V> withConsumerProperty(String configKey, Object configValue) {
+      return new ValueSourceBuilder<V>(underlying.withConsumerProperty(configKey, configValue));
+    }
+
+    public ValueSourceBuilder<V> withConsumerProperties(Map<String, Object> configToUpdate) {
+      return new ValueSourceBuilder<V>(underlying.withConsumerProperties(configToUpdate));
+    }
+
+    public ValueSourceBuilder<V> withValueCoder(Coder<V> valueCoder) {
+      return new ValueSourceBuilder<V>(underlying.withValueCoder(valueCoder));
+    }
+
+    public ValueSourceBuilder<V> withTimestampFn(SerializableFunction<V, Instant> timestampFn) {
+      return new ValueSourceBuilder<V>(
+          underlying.withTimestampFn(unwrapKafkaAndThen(timestampFn)));
+    }
+
+    /**
+     * A function to calculate watermark. When this is not set, last record timestamp is returned
+     * in {@link UnboundedReader#getWatermark()}.
+     *
+     * @param watermarkFn to calculate watermark at a record.
+     * @return Builder
+     */
+    public ValueSourceBuilder<V> withWatermarkFn(SerializableFunction<V, Instant> watermarkFn) {
+      return new ValueSourceBuilder<V>(
+          underlying.withTimestampFn(unwrapKafkaAndThen(watermarkFn)));
+    }
+
+    public UnboundedSource<V, KafkaCheckpointMark> build() {
+      return new UnboundedKafkaValueSource<V>((UnboundedKafkaSource<byte[], V>) underlying.build());
+    }
+
+    private static <InT, OutT>
+    SerializableFunction<KafkaRecord<byte[], InT>, OutT> unwrapKafkaAndThen(
+        final SerializableFunction<InT, OutT> fn) {
+      return new SerializableFunction<KafkaRecord<byte[], InT>, OutT>() {
+        public OutT apply(KafkaRecord<byte[], InT> record) {
+          return fn.apply(record.getValue());
+        }
+      };
+    }
+  }
+
+  /**
+   * Usually the users are only interested in value in KafkaRecord. This is a convenient class
+   * to strip out other fields in KafkaRecord returned by UnboundedKafkaValueSource
+   */
+  private static class UnboundedKafkaValueSource<V>
```

Comment:
Oh man. There has to be a way to make it possible that we only have a single implementation of the Source/Reader but we can support both with/without record metadata. I'll think and get back to you.